toolchain/parse/handle_decl_scope_loop.cpp

```diff
@@ -110,164 +257,17 @@ auto HandleDeclScopeLoop(Context& context) -> void {
-        if (TokenIsModifierOrIntroducer(
-                context.PositionKind(Lookahead::NextToken))) {
-          context.AddLeafNode(NodeKind::ExtendModifier, context.Consume());
-          saw_modifier = true;
-        } else {
-          // TODO: Treat this `extend` token as a declaration introducer
-          HandleUnrecognizedDecl(context, state.subtree_start);
-          return;
-        }
-        break;
-      }
-
-      // If we see a declaration introducer keyword token, replace the
-      // placeholder node and switch to a state to parse the rest of the
-      // declaration. We don't allow namespace or empty declarations here since
-      // they can't have modifiers and don't use bracketing parse nodes that
-      // would allow a variable number of modifier nodes.
-      case Lex::TokenKind::Class: {
-        introducer(NodeKind::ClassIntroducer,
-                   State::TypeAfterIntroducerAsClass);
-        return;
-      }
-      case Lex::TokenKind::Constraint: {
-        introducer(NodeKind::NamedConstraintIntroducer,
-                   State::TypeAfterIntroducerAsNamedConstraint);
-        return;
-      }
-      case Lex::TokenKind::Fn: {
-        introducer(NodeKind::FunctionIntroducer, State::FunctionIntroducer);
-        return;
-      }
-      case Lex::TokenKind::Interface: {
-        introducer(NodeKind::InterfaceIntroducer,
-                   State::TypeAfterIntroducerAsInterface);
-        return;
-      }
-      case Lex::TokenKind::Var: {
-        introducer(NodeKind::VariableIntroducer, State::VarAsDecl);
-        return;
-      }
-      case Lex::TokenKind::Let: {
-        introducer(NodeKind::LetIntroducer, State::Let);
-        return;
-      }
-
-      // We don't allow namespace or empty declarations after a modifier since
-      // they can't have modifiers and don't use bracketing parse nodes that
-      // would allow a variable number of modifier nodes.
-      case Lex::TokenKind::Namespace: {
-        introducer(NodeKind::NamespaceStart, State::Namespace);
-        return;
-      }
-      case Lex::TokenKind::Semi: {
-        if (saw_modifier) {
-          HandleUnrecognizedDecl(context, state.subtree_start);
-        } else {
-          context.ReplacePlaceholderNode(
-              state.subtree_start, NodeKind::EmptyDecl, context.Consume());
-        }
-        return;
-      }
-
-      default: {
-        // For anything else, report an error and output an invalid parse node
-        // or subtree.
-        HandleUnrecognizedDecl(context, state.subtree_start);
-        return;
-      }
+    auto position_kind = context.PositionKind();
+    if (TryHandleAsModifier(context, position_kind)) {
+      // Modifiers will continue through the loop, but we track we saw them.
+      saw_modifier = true;
+    } else if (TryHandleAsDecl(context, state, saw_modifier, position_kind)) {
+      return;
+    } else {
+      HandleUnrecognizedDecl(context, state.subtree_start);
+      return;
```

Comment:
I  think this would be a bit simpler if you moved the handling of the declaration out of the loop, eg:
```suggestion
  while (TryHandleAsModifier(context, context.PositionKind())) {
    saw_modifier = true;
  }
  if (!TryHandleAsDecl(context, state, saw_modifier, context.PositionKind())) {
    HandleUnrecognizedDecl(context, state.subtree_start);
  }
```
(Suggestion split in three parts to work around github limitations.)